[RC] WebSocket data gathering

domodwyer opened the pull request:
### What does this PR do?

Cause Agents that have Remote Config enabled to connect to the RC backend using a WebSocket connection every 24h (at `/api/v0.2/echo-test` - server side not yet implemented).

This connectivity test can be disabled by setting `remote_configuration.no_websocket_echo=true` in config. The test is opt-out due to the need to gather connectivity data from a representative sample of deployed Agents to evaluate the feasibility of switching to this new protocol in the future, or if it would cause significant breakage to existing users.

Every 24h a WebSocket connection is established, and the server sends a series of small data frames to the client, which immediately echos them back. It is not possible to use the WebSocket to induce any other behaviour in the Agent.

The implementation is careful to enforce safety limits (timeouts + payload size limits), avoids logging anything above DEBUG, and implements a panic recovery handler to minimise bug risk that might affect the production workload.

### Motivation

Gather connectivity / deliverability / latency / traffic management metrics, aiding development of a new RC / Agent protocol.

### Describe how you validated your changes

Tests are included for typical and edge case functionality.

### Possible Drawbacks / Trade-offs

The impact of this test should be negligible, and is designed to fail without causing any further effects.

### Additional Notes

Part of RC-2602.

This is my first PR to the Agent repo and I'm not really familiar with the codebase.

---

* feat(rc): WebSocket connectivity (9cd1980e6e)
      
      Add functionality to the api.HTTPClient to enable it to establish
      connections to a WebSocket service in the RC backend.
      
      Dialling a new WebSocket connection takes care to replicate the auth
      credential transmission mechanism used by the HTTP client, as well as
      the user-provided TLS & proxy config.

* feat(rc): implement WebSocket echo test (62657991e1)
      
      This commit implements a simple WebSocket echo test, connecting to the
      RC backend (at "/api/v0.2/echo-test") and exchanging a series of data
      frames.
      
      This test allows RC to evaluate the connectivity and performance of
      WebSockets in the wild, as a potential transport for a new protocol
      between the Agent and RC.
      
      Part of RC-2602.

* docs(rc): CoreAgentService.Stop() multiple calls (2a939cdcfb)
      
      Document that the Stop() function may be invoked multiple times: it is
      called >1 by at least some tests, even if not in the "prod" path.

* feat(rc): WebSocket connectivity tests every 24h (bd9f392e96)
      
      Make Remote Config enabled Agents perform periodic WebSocket
      connectivity tests every 24h.
      
      This connectivity test is required to evaluate real-world connectivity,
      through HTTP(S) proxies and egress filters, and gather latency and
      deliverability measurements that guide development of a new RC / Agent
      protocol.
      
      These tests are best effort and lightweight, exchanging a handful of
      frames with little memory or CPU overhead. They do not log any output
      above DEBUG level, except in the case of a panic which is caught to
      prevent the test affecting production workloads and logged at WARN.
      
      Once data gathering is complete, RC may stop servicing requests made to
      the test path ("/api/v0.2/echo-test") without causing any effects or
      logs on the Agent (tested).
      
      Tests can be explicitly disabled by configuring:
      
          remote_configuration.no_websocket_echo=true
      
      Part of RC-2602.

domodwyer committed:
feat(rc): WebSocket connectivity

Add functionality to the api.HTTPClient to enable it to establish
connections to a WebSocket service in the RC backend.

Dialling a new WebSocket connection takes care to replicate the auth
credential transmission mechanism used by the HTTP client, as well as
the user-provided TLS & proxy config.

domodwyer committed:
feat(rc): implement WebSocket echo test

This commit implements a simple WebSocket echo test, connecting to the
RC backend (at "/api/v0.2/echo-test") and exchanging a series of data
frames.

This test allows RC to evaluate the connectivity and performance of
WebSockets in the wild, as a potential transport for a new protocol
between the Agent and RC.

Part of RC-2602.

domodwyer committed:
docs(rc): CoreAgentService.Stop() multiple calls

Document that the Stop() function may be invoked multiple times: it is
called >1 by at least some tests, even if not in the "prod" path.

domodwyer committed:
feat(rc): WebSocket connectivity tests every 24h

Make Remote Config enabled Agents perform periodic WebSocket
connectivity tests every 24h.

This connectivity test is required to evaluate real-world connectivity,
through HTTP(S) proxies and egress filters, and gather latency and
deliverability measurements that guide development of a new RC / Agent
protocol.

These tests are best effort and lightweight, exchanging a handful of
frames with little memory or CPU overhead. They do not log any output
above DEBUG level, except in the case of a panic which is caught to
prevent the test affecting production workloads and logged at WARN.

Once data gathering is complete, RC may stop servicing requests made to
the test path ("/api/v0.2/echo-test") without causing any effects or
logs on the Agent (tested).

Tests can be explicitly disabled by configuring:

    remote_configuration.no_websocket_echo=true

Part of RC-2602.

domodwyer commented on pkg/config/remote/api/http.go:
```diff
@@ -256,6 +260,59 @@ func (c *HTTPClient) UpdateAPIKey(apiKey string) {
+// The "endpointPath" specifies the resource path to connect to, which is
+// appended to the client baseURL.
+func (c *HTTPClient) NewWebSocket(ctx context.Context, endpointPath string) (*websocket.Conn, error) {
+	// Extract the TLS & Proxy configuration from the HTTP client.
+	transport, ok := c.client.Transport.(*http.Transport)
+	if !ok {
+		return nil, errors.New("failed to extract transport config from HTTP client")
+	}
+
+	dialer := &websocket.Dialer{
+		HandshakeTimeout: 30 * time.Second,
+		TLSClientConfig:  transport.TLSClientConfig,
+		Proxy:            transport.Proxy,
+	}
```
WebSocket connections use the same TLS & proxy configuration as the RC HTTP client, which can be (and are) configured by the user in some deployments.

domodwyer commented on pkg/config/remote/api/http.go:
```diff
@@ -256,6 +260,59 @@ func (c *HTTPClient) UpdateAPIKey(apiKey string) {
+		Proxy:            transport.Proxy,
+	}
+
+	// The WebSocket request MUST include the same auth credentials as the plain
+	// HTTP requests.
+	c.headerLock.RLock()
+	headers := maps.Clone(c.header)
+	c.headerLock.RUnlock()
```
The WebSocket connections are authenticated in the same way as the RC HTTP requests.

domodwyer commented on pkg/config/remote/api/http_test.go:
```diff
@@ -0,0 +1,166 @@
+		serverHTTP2 bool
+		serverTLS   bool
+	}{
+		{
+			desc:        "tls with http1",
+			path:        "/",
+			serverHTTP2: false,
+			serverTLS:   true,
+		},
+		{
+			desc:        "tls with http2",
+			path:        "/",
+			serverHTTP2: true,
+			serverTLS:   true,
+		},
+		{
+			desc:        "plain with http1",
+			path:        "/",
+			serverHTTP2: false,
+			serverTLS:   false,
+		},
+		{
+			desc:        "plain with http2",
+			path:        "/",
+			serverHTTP2: true,
+			serverTLS:   false,
+		},
+		{
+			desc:        "path",
+			path:        "/api/v0.2/echo",
+			serverHTTP2: false,
+			serverTLS:   false,
+		},
```
Varying transport configurations are tested:

* TLS on/off
* HTTP vs HTTP2

This is probably going to be relevant in deployments that use non-transparent HTTP egress proxies irrespective of what RC uses in the backend.

domodwyer commented on pkg/config/remote/service/service.go:
```diff
@@ -471,6 +476,14 @@ func NewService(cfg model.Reader, rcType, baseRawURL, hostname string, tagsGette
 
 	clock := clock.New()
 
+	// WebSocket test actor - must call Start() to spawn the background task.
+	var websocketTest startstop.StartStoppable
+	if cfg.GetBool("remote_configuration.no_websocket_echo") {
+		websocketTest = &noOpRunnable{}
+	} else {
+		websocketTest = NewWebSocketTestActor(http)
+	}
```
My intention is to make it possible to disable this test, but I'm not familiar with how the config system works. 

Is this call sufficient to enable a user to disable the test or is there more config plumbing I need to do?

mellon85 replied:
Yeah, the only other alternative is adding a DD signed configurations to sync it to every org

domodwyer commented on pkg/config/remote/service/websocket_echo.go:
```diff
@@ -0,0 +1,98 @@
+//
+// This method is not concurrency safe, and panics if Start() has previously
+// been called.
+func (s *WebSocketTestActor) Start() {
+	if s.stopCh != nil {
+		panic("attempt to start WebSocketTestActor more than once")
+	}
+
+	s.stopCh = make(chan struct{})
+
+	go func() {
+		s.run()
+	}()
+}
```
I make the assumption that [the `CoreAgentService.Start()` method](https://github.com/DataDog/datadog-agent/blob/5624e49c09e450b27333fbb0b6969a763d951ac1/pkg/config/remote/service/service.go/#L542) is called once (and therefore concurrently).

I also assume that [the `CoreAgentService.Stop()` method](https://github.com/DataDog/datadog-agent/blob/5624e49c09e450b27333fbb0b6969a763d951ac1/pkg/config/remote/service/service.go/#L639) _can_ be called more than once (at least one test does) but never concurrently.

Are these assumptions correct?

domodwyer reviewed (commented):
I've left some review notes and questions :+1:

mellon85 commented on comp/remote-config/rcwebsocket/echo.go:
```diff
@@ -0,0 +1,126 @@
+	)
+
+	// Perform the frame echo test routine.
+	n := uint(0)
```
```suggestion
	attemptCount := uint(0)
```

domodwyer committed:
chore: add copyright headers

michaelcretzman commented on releasenotes/notes/RC-WebSocket-connectivity-tests-9f04946537a732ad.yaml:
```diff
@@ -0,0 +1,40 @@
+# Each section from every release note are combined when the
```
looks like this is still in Draft as there is only placeholder text here.

domodwyer committed:
refactor(rc): move packages for cmd -> comp -> pkg

Imports should be cmd -> comp -> pkg, and the service init is in pkg,
meaning any code that it references should be in pkg too.

domodwyer committed:
refactor: remove release notes

After a discussion with RC manager, it was decided this didn't need
explicit release notes.

domodwyer committed:
build: run dda inv tidy

domodwyer committed:
refactor: fix lint failures

domodwyer committed:
test: fix racy stop in tests

domodwyer committed:
chore: update LICENSE-3rdparty

I'm manually editing this based on what CI spat out.
